src/ln/channelmanager.rs

```diff
@@ -2145,29 +2231,27 @@ impl ChannelManager {
 											}
 										}
 									}
+									return_node_fail_update!(res, route_hop.pubkey, true);
 								},
-								_c if error_code & BADONION == BADONION => {
-									//TODO
-								},
-								14 => { // expiry_too_soon
+								21 => {
+									onion_failure_log!("expiry_too_far", error_code);
 									res = Some((None, true));
 									return;
-								}
+								},
 								_ => {
-									// node sending unknown code
-									res = Some((Some(msgs::HTLCFailChannelUpdate::NodeFailure {
-										node_id: route_hop.pubkey,
-										is_permanent: true,
-									}), false));
-									return;
+									return_node_fail_update!(res, route_hop.pubkey, true);
 								}
 							}
 						}
 					}
 				}
 			}).expect("Route that we sent via spontaneously grew invalid keys in the middle of it?");
-			res.unwrap_or((None, true))
-		} else { ((None, true)) }
+			let (channel_update, payment_retryable) = res.expect("should have been set!");
```

Comment:
If a node simply garbles the onion data we'll panic at this expect - we have to just move on since in such a case (presumably retrying the payment, since its likely an intermediate node garbled it) as we have no ability to know what happened.